_source/_posts/2021-10-16-full-stack-java.md

```diff
@@ -0,0 +1,727 @@
+
+To switch from Keycloak to Auth0, you only need to override the default properties (for Spring Security OAuth). You don't even need to write any code!
+
+To see how it works, create a `.auth0.env` file in the root of your project, and fill it with the code below to override the default OIDC settings:
+
+```shell
+export SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_OIDC_ISSUER_URI=https://<your-auth0-domain>/
+export SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_ID=<your-client-id>
+export SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_SECRET=<your-client-secret>
+export JHIPSTER_SECURITY_OAUTH2_AUDIENCE=https://<your-auth0-domain>/api/v2/
+```
+
+**WARNING:** Modify your existing `.gitignore` file to have `*.env` so you don't accidentally check in your secrets!
+
+You'll need to create a new web application in Auth0 and fill in the `<...>` placeholders before this works. 
+
+### Create an OpenID Connect App on Auth0
+
+Log in to your Auth0 account (or [sign up](https://auth0.com/signup) if you don't have an account). You should have a unique domain like `dev-xxx.eu.auth0.com`. 
+
+Press the **Create Application** button in [Applications section](https://manage.auth0.com/#/applications). Use a name like `JHipster Baby!`, select `Regular Web Applications`, and click **Create**.
+
+Switch to the **Settings** tab and configure your application settings:
+
+- Allowed Callback URLs: `http://localhost:8080/login/oauth2/code/oidc`
+- Allowed Logout URLs: `http://localhost:8080/`
+
+Scroll to the bottom and click **Save Changes**.
+
+In the [roles](https://manage.auth0.com/#/roles) section, create new roles named `ROLE_ADMIN` and `ROLE_USER`.
+
+Create a new user account in the [users](https://manage.auth0.com/#/users) section. Click on the **Role** tab to assign the roles you just created to the new account. 
+
+_Make sure your new user's email is verified before attempting to log in!_
+
+Next, head to **Auth Pipeline** > **Rules** > **Create**. Select the `Empty rule` template. Provide a meaningful name like `Group claims` and replace the Script content with the following.
+
+```js
+function(user, context, callback) {
+  user.preferred_username = user.email;
+  const roles = (context.authorization || {}).roles;
+
+  function prepareCustomClaimKey(claim) {
+    return `https://www.jhipster.tech/${claim}`;
+  }
+
+  const rolesClaim = prepareCustomClaimKey('roles');
+
+  if (context.idToken) {
+    context.idToken[rolesClaim] = roles;
+  }
+
+  if (context.accessToken) {
+    context.accessToken[rolesClaim] = roles;
+  }
+
+  callback(null, user, context);
+}
+```
+
+This code is adding the user's roles to a custom claim (prefixed with `https://www.jhipster.tech/roles`). This claim is mapped to Spring Security authorities in `SecurityUtils.java`.
+
+```java
+public static List<GrantedAuthority> extractAuthorityFromClaims(Map<String, Object> claims) {
+    return mapRolesToGrantedAuthorities(getRolesFromClaims(claims));
```

Comment:
if you're showing this extract method, I think you should also show mapRolesTo... and getRolesFrom.. methods so that everything can be read from that code block